Create/Edit image secret

jhadvig opened the pull request:
Adding support for creating image pull secrets:
- kubernetes.io/dockerconfigjson - new format
```json
{
	"auths": {
		"https://index.docker.io/v1/": {
			"auth": "dGVzdHVzZXIzOnRlc3RwdzM=",
			"email": "jhadvig@test.com"
		}
	}
}
```

- kubernetes.io/dockercfg  - old format
```json
{
	"https://index.docker.io/v1/": {
		"auth": "dGVzdHVzZXIzOnRlc3RwdzM=",
		"email": "jhadvig@test.com"
	}
}
```

The secret contains credentials for 1-n registry entries. Each secret has to contain `auth` field, which is base64 encoded string, that encodes `${username}:${password}` string, or both `username` and `password` themselves. Also `mail` has to be there

User can either create a secret by providing the registry credentials in the `CreateConfigSubform` or upload configuration file in `UploadConfigSubform`.

![1](https://user-images.githubusercontent.com/1668218/42329745-be1db3ee-8071-11e8-94d4-0bec7ea37426.gif)

![2](https://user-images.githubusercontent.com/1668218/42329746-be451d12-8071-11e8-9d64-b7a7786aeb95.gif)

![3](https://user-images.githubusercontent.com/1668218/42329747-be62a3f0-8071-11e8-8ebc-6dd385e1f391.gif)

Editing of the image secret is for now done via `CreateConfigSubform` form and its possible to edit just the first registry credentials. Wasn't really sure how to approach this issue image secret can containe multiple registry credentials. There are couple of ideas how to tackle this issue:
- do the editation of the secret purely in the textarea of the `UploadConfigSubform` component.
   - cons: not the best UX + could cause some unintentional mistakes.
- have a `server address` picker above the `CreateConfigSubform` by which user can select and edit desired secret
   - cons: the `server address` should be also editable + not sure how server entry should be deleted then
- Just list all the registry entries using `CreateConfigSubform` component and will also be able to delete the particular entry
   - cons: list can get long

@spadgett @benjaminapetersen @openshift/team-ux-review PTAL

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -139,6 +160,211 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+        loadedData.username = serverCredentials[serverAddress].username;
+        loadedData.password = serverCredentials[serverAddress].password;
+      } else {
+        const auth = window.atob(_.get(serverCredentials, [serverAddress, 'auth'])).split(':');
```
Quick example about why I'd like to see some smaller named functions.  I'm not entirely sure why we `const auth` in the `else` branch of a second `if` block.  I'm sure there is reasoning in the flow, and if I think about it a bit longer I can probably figure it out, but it would be better to describe the "why" via a named function call.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -139,6 +160,211 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+        <FileInput onChange={this.onFileChange} />
+        <p className="help-block">Upload a .dockercfg or .docker/config.json file</p>
+        <textarea className="form-control co-create-secret-form__textarea"
+          id="docker-config"
```
I haven't looked at this a ton in this console, but in the old `angular` console we would occasionally have errors due to duplicate `id` attribs.  Should this generate a unique ID each time the form is rendered to the page?  Do we care that the `name` and `id` are following different conventions?

spadgett replied:
Shouldn't be an issue unless the form is rendered twice on the same page. I'd rather not add the logic until we need it.

benjaminapetersen replied:
I wouldn't hold a PR over it, but I tend to think that if we use an `id` we should always ensure it is unique. I'm not sure it hurts anything, and the logic is pretty minimal via `_.uniqueId('prefix-')`.  I can see where it may get a touch more complicated if the `id` needs to be created in `componentWillMount()` & stored via `setState()`.  

No worries here though since it isn't standard.

spadgett replied:
It's just not needed on standalone pages and adds complexity. It's also easy to get wrong when you start generating dynamic values and can make test code harder to write.

beanh66 commented:
@jhadvig I can provide a more detailed design in the morning! I think allowing the user to edit the same way they created makes sense. So if they created using file upload, I would expect the file to be editable and if created via the form, editing would display a form.

jhadvig commented:
@beanh66 I don't think that we can actually know how the image secret was created. The thing is that via the form you can create image secret with one registry entry, eg:
```json
{
	"auths": {
		"https://index.docker.io/v1/": {
			"auth": "dGVzdHVzZXIzOnRlc3RwdzM=",
			"email": "jhadvig@test.com"
		}
	}
}
```

With the text area you can create image secret with 1-n registry entries, eg:
```json
{
	"auths": {
		"https://index.docker.io/v1/": {
			"auth": "dGVzdHVzZXIzOnRlc3RwdzM=",
			"email": "jhadvig@test.com"
		},
		"https://index.disneyland.io/v1/": {
			"username": "jhadvig",
			"password": "jakub",
			"auth": "dGVzdHVzZXIzOnRlc3RwdzM=",
			"email": "jhadvig@test.com"
		}
	}
}
```

but we  are not actually storing any meta about how its created. So in that case we could edit secret with one entry via form and 2-n entries via text area. The thing about the text area is that in case the user will change the `username` or `password`, we would need to take update the `auth` field with the new value, since the `auth`value is base64 encoded "{username}:${password}", or the user will have to do it himself.

spadgett commented on frontend/public/components/secret.jsx:
```diff
@@ -21,11 +21,16 @@ export const WebHookSecretKey = 'WebHookSecretKey';
+//     - kubernetes.io/dockercfg
 // - webhook secret with one key.
 const editInYaml = obj => {
   switch (obj.type) {
```
It's not brittle when the case clauses are empty and they're stacked like this. In fact, eslint specifically allows this, but will complain if you have any logic between in the case clause then fall through.

benjaminapetersen replied:
Agree, that is acceptable enough, I'd strongly assert change if there was logic.

However, I'd still like to see the comments expressed in code, if possible.

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -139,6 +160,211 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+    }
+
+    this.state = {
+      isNewFormat: _.isEmpty(loadedData) || Object.keys(this.props.stringData)[0] === AUTHS_SYMBOL,
```
```js
_.isEmpty(loadedData) || !!this.props.stringData[AUTHS_SYMBOL]
```

Could use a comment explaining why empty data means we use the new format.

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -139,6 +160,211 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+  render() {
+    return <React.Fragment>
+      <div className="form-group">
+        <label className="control-label">Registry Server Address</label>
```
htmlFor="address"

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -139,6 +160,211 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+        </div>
+      </div>
+      <div className="form-group">
+        <label className="control-label">Username</label>
```
htmlFor="username"

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -139,6 +160,211 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+        </div>
+      </div>
+      <div className="form-group">
+        <label className="control-label">Email</label>
```
htmlFor="email"

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -139,6 +160,211 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+  constructor (props) {
+    super(props);
+    this.state = {
+      configFile: _.isEmpty(this.props.stringData) ? '' : JSON.stringify(this.props.stringData),
```
I'm confused here... Isn't the config file one value of one key? Why is it set to all of stringData?

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -139,6 +160,211 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+      <label className="control-label" htmlFor="docker-config">Configuration File</label>
+      <div>
+        <FileInput onChange={this.onFileChange} />
+        <p className="help-block">Upload a .dockercfg or .docker/config.json file</p>
```
nit: end the sentence with a period

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -312,7 +545,10 @@ class SSHAuthSubform extends React.Component<SSHAuthSubformProps, SSHAuthSubform
+  if (!props.obj.loaded) {
+    return <LoadingBox />;
+  }
+  const secretTypeAbstraction = determineSecretTypeAbstraction(_.get(props.obj, ['data', 'type']));
```
_.get(props.obj, 'data.type')

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -360,6 +596,41 @@ export type BaseEditSecretProps_ = {
 
+export type ImageSecretFormState = {
+  type: SecretType,
+  stringData: any,
```
`[key: string]: any` as below?

spadgett replied:
Actually these should all be

```ts
[key: string]: string
```

right?

cshinn commented:
@jhadvig I think it makes sense to do form editing for a single entry and show the config editor for multiple entries. As for the base64 encoding, would it be possible for us to change that after submission or something so that they don't need to do it manually?

jhadvig commented:
@cshinn well we could replace decode the new `auth` value from updated `username` and `password` and replace the old one.

spadgett commented:
/hold
for feature freeze

beanh66 commented:
@jhadvig here is a recording of a design option: http://recordit.co/vLSjAoBl22. The important change would be to include the Add/Remove of additional credentials: 
![secrets-1](https://user-images.githubusercontent.com/18057044/43002660-8fee76f2-8bf7-11e8-8612-adb03b5c0cca.png)
![secrets-2](https://user-images.githubusercontent.com/18057044/43002661-9021f270-8bf7-11e8-86f3-bd0eeee50f9b.png)

@openshift/team-ux-review We need a way to show the credentials when users edit a secret, but we are unable to tell how the secret was created in the first place. Per my discussion with @jhadvig, adding the capability to show more than one credential in the form would allow us to show edit mode using the form, regardless of how the secret was created. We could now account for multiple credentials that were possibly added using a json file.

jhadvig commented:
@spadgett @benjaminapetersen PR updated, based on the comments.
The PR is based on the [Drag&Drop for FileInput](https://github.com/openshift/console/pull/174) PR since the file uploading.
For creating the image secret user can pick from:
- filling up the form and creating image secret with single entry
- uploading the config file

When editing the image secret a image secret form will be shown with as many `ConfigEntryForm`'s as many entries in the secret.
![1](https://user-images.githubusercontent.com/1668218/43325282-8d918722-91b6-11e8-8a5f-249d51ccbb03.png)


Will add the logic for adding and removing entries in a followUp.
I know that the commit is a bit bigger but I've added additional commits to make it more straight forward.
PTAL

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -307,11 +628,21 @@ class SSHAuthSubform extends React.Component<SSHAuthSubformProps, SSHAuthSubform
 }
 
 const secretFormFactory = secretType => {
```
Existing code, but better to add types to the signature

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -307,11 +628,21 @@ class SSHAuthSubform extends React.Component<SSHAuthSubformProps, SSHAuthSubform
+  if (!props.obj.loaded) {
+    return <LoadingBox />;
+  }
+  const secretTypeAbstraction = determineSecretTypeAbstraction(_.get(props.obj, 'data.type'));
```
Not sure if `_.get` is needed here if we're checking if it's loaded already?

jhadvig replied:
yeah, its not necessary here :+1:

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -367,6 +698,43 @@ export type BasicAuthSubformProps = {
+export type ImageSecretFormState = {
+  type: SecretType,
+  stringData: {
+    [key: string]: any,
```
Shouldn't this be (here and below)

```ts
[key: string]: string,
```

Probably worth making `stringData` a separate type you can reuse

jhadvig replied:
for the `ImageSecretFormProps` it makes sense to have it `{[key: string]: string}`, not sure whats the right type for `ImageSecretFormState`, since the `stringData` looks like: 
```json
{  
   ".dockerconfigjson":{  
      "auths":{  
         "https://index.docker.io/v1/":{  
            "username":"testuser",
            "password":"testpw",
            "auth":"dGVzdHVzZXI6dGVzdHB3",
            "email":"jhadvig@test.com"
         },
         "https://index.docker.io/v2/":{  
            "username":"testuser",
            "password":"testpw",
            "auth":"dGVzdHVzZXI6dGVzdHB3",
            "email":"jhadvig2@test.com"
         }
      }
   }
}
```

tried `{[key: string]: Object}` but that was throwing type errors:
```
TS2326: Types of property 'stringData' are incompatible.
  Type 'Object' is not assignable to type '{ [key: string]: Object; }'.
    The 'Object' type is assignable to very few other types. Did you mean to use the 'any' type instead?
      Index signature is missing in type 'Object'.
```

One option would be to create a separate types and use those, but we will need probably two of them since `.dockercfg` and `.dockerconfigjson` have different structure. Not sure if that's worth if we can use `any`

spadgett replied:
Ah, OK. If it's an actual object and not the serialized string, then `any` is OK

spadgett commented on frontend/public/components/secrets/_create-secret.scss:
```diff
@@ -0,0 +1,13 @@
+  margin-bottom: 25px;
+}
+
+.co-create-image-secret__form:not(:first-of-type) {
```
If these are siblings, this could be

```css
.co-create-image-secret__form + .co-create-image-secret__form {
```

jhadvig replied:
Nicee ! :+1:

jhadvig commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -33,16 +35,33 @@ export type BasicAuthSubformState = {
-  return typeAbstraction === SecretTypeAbstraction.source ? SecretType.basicAuth : SecretType.opaque;
+  switch (typeAbstraction) {
+    case SecretTypeAbstraction.source:
+      return SecretType.basic;
```
Makes sense to keep the '*Auth' for the 'kubernetes.io/basic-auth' && 'kubernetes.io/ssh-auth'

jhadvig commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -144,6 +163,308 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
   }
 };
 
+const getImageSecretKey = (secretType) => {
```
Hmm... both of these functions `getImageSecretKey` and  `getImageSecretType` have just two keys or types to return, since they are used for the image-secret types, which are just 'kubernetes.io/dockerconfigjson and 'kubernetes.io/dockercfg', so I this that simple `if` statement with defaulting to the other type/key should be sufficient.

jhadvig replied:
Or do we want to default to `Opaque` SecretType ? So we have control over the returned types ?

spadgett replied:
I think the implementation is fine, just that we should add the proper types to the function signature to take advantage of TypeScript

jhadvig commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -144,6 +163,308 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+        return false;
+      }
+    });
+    return hasDuplicate;
```
Make sense !

jhadvig commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -144,6 +163,308 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+    return (
+      <React.Fragment>
+        {secretEntriesList}
+        { this.state.hasDuplicate && <div className="co-create-secret-warning">Updated <b>Registry Server Address</b> is duplicate. Remove it, or it will be removed upon saving.</div> }
```
Yeah, I wasn't sure if we want to check for the dup. here, since it doesn't make sense to have different credentials pointing to the same address in one secret. 
Other think is if the users know that :-/

jhadvig commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -144,6 +163,308 @@ const withSecretForm = (SubForm) => class SecretFormComponent extends React.Comp
+    this.setState({
+      stringData: {[this.state.dataKey]: secretData}
+    }, () => this.props.onChange({
+      stringData: _.mapValues(this.state.stringData, JSON.stringify),
```
Hmm not sure if an exception can occure here, since the underlying components - `CreateConfigSubform` and `UploadConfigSubform` should take care of it.

In `CreateConfigSubform` when an error occurs over parsing a JSON, the editor should disable the `Save` button (at least thats the logic Im addressing in the upcoming commit)
![1](https://user-images.githubusercontent.com/1668218/43393530-cf4c07e4-93f7-11e8-9782-dea528ec1d43.png)


For the `UploadConfigSubform` - upon secret creation, when the user pastes/enters a non-JSON format to the text-area  he will be notified with a warning:
![2](https://user-images.githubusercontent.com/1668218/43393532-d17b034e-93f7-11e8-89ec-0b41d8e194b0.png)
Here the question is, whether we want to disable the 'Create' button or have it still clickable but don't propagate the changes to the upper components (as the current code is doing).

spadgett replied:
I need to look at this more closely again, but generally speaking: as much as possible, I'd like the console to be a 
"dumb" client that just shows the value and sends what the user types in and lets the API server do any validation / use the data. Problems then are communicated to the user through either API errors on save or events and conditions later on.

Specifically in this case, do we know that this is always going to be a JSON file? What if someday YAML is accepted as a config file format?

I want to be careful of starting down the road of reimplementing logic from API server. We went too far in that direction in the last console, and it causes problems as the API server changes and implementations drift. Even if it works perfectly now, too many assumptions about the format can cause bugs later on.

It might make sense to separate the cases where we create the config file based on individual inputs vs editing an existing value or handling a file upload. If it's not a file we created, we just treat it as a blob.

jhadvig replied:
Having a 'dumb' client makes sense together with communicating the errors through API.
But this case is slightly different. If we take the issue of creating the pull/push secret using CLI (only other user access point, besides UI console), there are two ways how to create a pull/push secret:
1. using `oc create secret docker-registry`, where user has to specify - username & pw & email + server(optional). This call will create only the new `k8s.io/dockerconfigjson` secret type.

2. using `oc create secret generic pull --from-file=.dockercfg=../origin/config.json --type=kubernetes.io/dockercfg` which user has to feed with config file (dockercfg or dockerconfigjson format) but he also has to specify the right type of the format.

The first case we use when user fills the form with the server& username & pw & email inputs.

The second one is a bit tricky cause here the user has to know the format type, to help the CLI with creating the right secret type, since its not parsing the config file to find out whats the type. Here I wasnt sure how to do it, without parsing the file.

One thing that comes to my mind is to have a additional selectbox above the config file droparea where user can specify the secret type, `k8s.io/dockercofigjson` by default. That case, the experience would be similar to the one from the CLI, and we could threat the file input as a blob and just wait for the API server response, if something is wrong. This way we can skip any kind of validation.
Thoughts ?

benjaminapetersen commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -1,50 +1,153 @@
+                  disabled />
+                <span className="input-group-btn">
+                  <span className="btn btn-default co-btn-file">
+                    Browse&hellip;
```
Assuming `Browse&hellip;` is ellipsis & not an oops :)

jhadvig replied:
`&hellip;` stands for **H**orizontal **Ellipsis**, that's an [HTML entity](https://www.toptal.com/designers/htmlarrows/punctuation/horizontal-ellipsis/). It looks like this: `…`

jhadvig commented:
@spadgett I've addressed the comments.
Now when the uploaded config file won't be in the JSON, the Save/Create button will be disabled, for that I've `onFormDisable` helper that will check the state of the uploaded JSON.

spadgett commented:
/hold cancel

jhadvig commented:
@spadgett I've rebased the PR. Kept the additional commit with changes.
PTAL

spadgett commented:
@jhadvig I notice if you add random JSON using file upload and then edit then edit the secret later, the form doesn't handle it well (although it doesn't error). I'm not sure the best way to handle that, but probably worth investigating.

spadgett commented:
/lgtm
/hold cancel

spadgett commented:
jenkins rebuild

spadgett commented:
jenkins rebuild

spadgett commented:
jenkins rebuild

spadgett commented:
jenkins rebuild
